injectproxy/routes.go

```diff
@@ -1,47 +1,93 @@
 package injectproxy
 
 import (
+	"context"
 	"fmt"
 	"net/http"
+	"net/http/httputil"
+	"net/url"
 
 	"github.com/prometheus/prometheus/pkg/labels"
 	"github.com/prometheus/prometheus/promql"
 )
 
 type routes struct {
-	handler http.Handler
-	label   string
+	handler    http.Handler
+	label      string
+	forwarders map[string]func(http.ResponseWriter, *http.Request)
+	modifiers  map[string]func(*http.Response) error
 }
 
-func NewRoutes(handler http.Handler, label string) *routes {
-	return &routes{
-		handler: handler,
+func NewRoutes(upstream *url.URL, label string) *routes {
+	proxy := httputil.NewSingleHostReverseProxy(upstream)
+
+	r := &routes{
+		handler: proxy,
 		label:   label,
 	}
+	r.forwarders = map[string]func(http.ResponseWriter, *http.Request){
+		"/federate":           r.federate,
+		"/api/v1/query":       r.query,
+		"/api/v1/query_range": r.query,
+		"/api/v1/rules":       r.noop,
+	}
+	r.modifiers = map[string]func(*http.Response) error{
+		"/api/v1/rules": r.rules,
+	}
+	proxy.ModifyResponse = r.ModifyResponse
+	return r
 }
 
 func (r *routes) ServeHTTP(w http.ResponseWriter, req *http.Request) {
-	if req.URL.Path == "/api/v1/query" || req.URL.Path == "/api/v1/query_range" {
-		r.query(w, req)
+	h, found := r.forwarders[req.URL.Path]
+	if !found {
+		http.NotFound(w, req)
 		return
 	}
-	if req.URL.Path == "/federate" {
-		r.federate(w, req)
+
+	lvalue := req.URL.Query().Get(r.label)
+	if lvalue == "" {
+		http.Error(w, fmt.Sprintf("Bad request. The %q query parameter must be provided.", r.label), http.StatusBadRequest)
 		return
 	}
+	req = req.WithContext(withLabelValue(req.Context(), lvalue))
+	req.URL.Query().Del(r.label)
 
-	http.NotFound(w, req)
+	h(w, req)
 }
 
-func (r *routes) query(w http.ResponseWriter, req *http.Request) {
-	q := req.URL.Query()
-	labelValue := q.Get(r.label)
-	if labelValue == "" {
-		http.Error(w, fmt.Sprintf("Bad request. The %q query parameter must be provided.", r.label), http.StatusBadRequest)
-		return
+func (r *routes) ModifyResponse(resp *http.Response) error {
+	m, found := r.modifiers[resp.Request.URL.Path]
+	if !found {
+		return nil
```

Comment:
we are not explicitly setting 404 here. Looking at https://golang.org/pkg/net/http/httputil/#ReverseProxy I don't see explicit behavior in case of a `nil` return. Can we reference/document what happens from here with the response?

Reply:
Sure i'll add this in a following PR. AFAIU when `ModifyResponse` doesn't return an error then the response is forwarded as-is to the client.